### PR TITLE
Fix packages shipping NuGet's placeholder description and guard package metadata

### LINKED ISSUE #10368: Repository Quality: NuGet Package Discoverability and Metadata Completeness
### 🎯 Repository Quality Improvement Report — NuGet Package Discoverability and Metadata Completeness

**Analysis Date**: 2026-07-31
**Focus Area**: NuGet Package Discoverability and Metadata Completeness
**Strategy Type**: Custom

### Executive Summary

This repository ships **21 NuGet packages** spanning MSTest, Microsoft.Testing.Platform, and 15+ MTP extensions. Every single one of them is missing a `PackageReadmeFile` entry — the NuGet.org feature that renders a rich README on the package detail page. Today, visitors to nuget.org/packages/MSTest or nuget.org/packages/Microsoft.Testing.Platform see only the plain-text `PackageDescription` boilerplate, which provides a poor discovery and documentation experience.

The MSTest.Sdk package is a particularly missed opportunity: it already has a high-quality `README.md` co-located in `src/Package/MSTest.Sdk/`, but the file is **not wired up** to `PackageReadmeFile` in the `.csproj`, so it is never embedded in the `.nupkg` and never shown on nuget.org.

One package — `Microsoft.Testing.Extensions.AzureFoundry` — has **no `PackageDescription` at all**, while several MTP extension descriptions reuse generic text that does not differentiate what the package specifically provides.

<details>
<summary><b>Full Analysis Report</b></summary>

### Current State Assessment

**Metrics Collected:**

| Metric | Value | Status |
|--------|-------|--------|
| Total shipped NuGet packages | 21 | — |
| Packages with `PackageReadmeFile` set | 0 / 21 | ❌ |
| Packages with `PackageDescription` set | 20 / 21 | ⚠️ |
| Packages with fully boilerplate generic description | ~8 (MTP extensions) | ⚠️ |
| `MSTest.Sdk/README.md` exists but not wired | 1 | ❌ |
| Packages with `PackageIcon` set (centrally or locally) | 0 verified | ⚠️ |

**Packages without `PackageReadmeFile`** (all 21):
- MSTest.TestFramework, MSTest.Analyzers, MSTest.SourceGeneration, MSTest, MSTest.Sdk, MSTest.TestAdapter
- Microsoft.Testing.Platform
- Microsoft.Testing.Extensions.CrashDump, .HangDump, .TrxReport, .HtmlReport, .JUnitReport, .CtrfReport, .Retry, .HotReload, .Telemetry, .VSTestBridge, .GitHubActionsReport, .AzureDevOpsReport, .VideoRecorder, .PackagedApp

**Package with missing description:**
- `Microsoft.Testing.Extensions.AzureFoundry` — no `<PackageDescription>` element at all.

**Packages with unchanged boilerplate opening line** (`"Microsoft Testing is a set of platform, framework and protocol intended to make it possible to run any test on any target or device."`):
CrashDump, Platform, Telemetry, GitHubActionsReport, AzureDevOpsReport, HangDump, JUnitReport, CtrfReport, HtmlReport, VideoRecorder, TrxReport, Retry (12 packages share the identical first paragraph).

#### Strengths
- `MSTest.Sdk` already has a co-located `README.md` with tables, code blocks, and structured sections — exactly what nuget.org can render.
- MTP extension descriptions are specific in their **second** paragraph (e.g., "This package extends Microsoft Testing Platform to provide TRX test reports.") — the per-package differentiation text already exists.
- `PackageReleaseNotes` is centrally set in `src/Directory.Build.props` and `src/Platform/Directory.Build.props` to link to the changelogs — a good discoverability signal already in place.

#### Areas for Improvement
- **Critical — zero packages use `PackageReadmeFile`**: nuget.org shows no rendered README for any of the 21 packages.
- **High — `MSTest.Sdk` README exists but is never embedded**: `src/Package/MSTest.Sdk/README.md` exists locally but is not included via `PackageReadmeFile` in the `.csproj`.
- **Medium — `AzureFoundry` missing description entirely**: alpha package, but still ships without `<PackageDescription>`.
- **Low — generic opening paragraph duplicated across 12 MTP extension descriptions**: each extension starts with an identical sentence that adds no value for package searchers.

</details>

---

### 🤖 Suggested Improvement Tasks

The following actionable tasks address the findings above.

#### Task 1: Wire up the existing `MSTest.Sdk/README.md` as `PackageReadmeFile`

**Priority**: High
**Estimated Effort**: Small

`src/Package/MSTest.Sdk/README.md` is already co-located with the project but is never included in the `.nupkg`. Add the `PackageReadmeFile` property and an `<None>` item to pack it:

```xml
<!-- In src/Package/MSTest.Sdk/MSTest.Sdk.csproj -->
<PropertyGroup>
  <PackageReadmeFile>README.md</PackageReadmeFile>
</PropertyGroup>
<ItemGroup>
  <None Include="README.md" Pack="true" PackagePath="\" />
</ItemGroup>
```

This is a zero-content-writing task — the file already exists with high-quality content.

---

#### Task 2: Create and wire `PackageReadmeFile` READMEs for the core MSTest packages

**Priority**: High
**Estimated Effort**: Medium

Create `README.md` files for the six core MSTest packages and configure them via `PackageReadmeFile`:

- `src/TestFramework/TestFramework.Extensions/` → `MSTest.TestFramework` — describe the attributes, Assert, TestContext, data-driven tests
- `src/Adapter/MSTest.TestAdapter/` → `MSTest.TestAdapter` — describe discovery and execution, integration with test hosts
- `src/Package/MSTest/` → `MSTest` — describe the meta-package, quick-start snippet
- `src/Analyzers/MSTest.Analyzers.Package/` → `MSTest.Analyzers` — describe code-fix coverage, how to configure rules
- `src/Analyzers/MSTest.SourceGeneration/` → `MSTest.SourceGeneration` — describe reflection-free mode, NativeAOT, opt-in mechanics

Consider adding a shared `PackageReadmeFile` default to `src/Directory.Build.props` once the files exist.

---

#### Task 3: Create `PackageReadmeFile` READMEs for the Microsoft.Testing.Platform and its extensions

**Priority**: Medium
**Estimated Effort**: Medium

Create per-package README files for `Microsoft.Testing.Platform` and the 14 shipping MTP extensions. Each file needs only:
1. A one-paragraph summary of what the package provides
2. A quick-start code snippet showing how to register the extension
3. A link to the relevant documentation section in `docs/`

The documentation for many extensions already exists under `docs/microsoft.testing.platform/`. Consider adding a shared default in `src/Platform/Directory.Build.props`.

---

#### Task 4: Add `PackageDescription` to `Microsoft.Testing.Extensions.AzureFoundry`

**Priority**: Medium
**Estimated Effort**: Small

`src/Platform/Microsoft.Testing.Extensions.AzureFoundry/Microsoft.Testing.Extensions.AzureFoundry.csproj` has no `<PackageDescription>` element. Even though the package is currently alpha/pre-release (`IsShipping=false`), it will ship eventually and the description will be missing when it does. Add a meaningful description explaining the Azure Foundry / AI integration use-case.

---

#### Task 5: Reduce boilerplate in MTP extension descriptions to improve nuget.org search relevance

**Priority**: Low
**Estimated Effort**: Small

Twelve MTP extension packages share the identical opening line `"Microsoft Testing is a set of platform, framework and protocol intended to make it possible to run any test on any target or device."` This adds noise and dilutes the specific value proposition of each package in nuget.org search results. Either:
- Move the generic sentence to a `PackageReadmeFile` preamble and make each `PackageDescription` lead with the specific purpose (e.g., "Extends Microsoft Testing Platform to generate TRX test reports."), **or**
- Centralize the generic preamble in `src/Platform/Directory.Build.props` using a property and only require each package to set the unique suffix.

---

### 📊 Historical Context

<details>
<summary><b>Previous Focus Areas</b></summary>

| Date | Focus Area | Type |
|------|------------|------|
| 2026-07-30 | assert-interpolated-string-handler-coverage-gap | Custom |
| 2026-07-29 | warning-suppression-technical-debt | Custom |
| 2026-07-23 | parameterized-test-display-name-ux | Custom |
| 2026-07-21 | analyzer-code-fix-actionability-gap | Custom |
| 2026-07-20 | sourcegen-reflectionfree-coverage-gaps | Custom |
| 2026-07-16 | mstest-sdk-msbuild-property-reference-gap | Custom |
| 2026-07-14 | assertion-telemetry-coverage-gap | Custom |
| 2026-07-10 | cicd-workflow-hygiene | Standard |
| 2026-07-09 | extension-async-cancellation-completeness | Custom |
| 2026-07-08 | extension-uid-stability-governance | Custom |

</details>

---

#### Immediate Actions (This Week)
1. Wire `MSTest.Sdk/README.md` to `PackageReadmeFile` — Priority: High (zero content writing needed)
2. Add `PackageDescription` to `AzureFoundry` — Priority: Medium (one-liner fix)

#### Short-term Actions (This Month)
1. Author and wire README files for the six core MSTest packages — Priority: High
2. Author and wire README files for Microsoft.Testing.Platform and its shipping extensions — Priority: Medium
3. Refactor boilerplate MTP extension descriptions — Priority: Low

*Next analysis: 2026-08-01 — Focus area selected based on diversity algorithm*

> 🤖 **Automated content by GitHub Copilot.** Generated by the [Repository Quality Improver](https://github.com/microsoft/testfx/actions/runs/30670197583/agentic_workflow) workflow. · sonnet46 69 AIC · ⌖ 7.94 AIC · ⊞ 10.8K · [◷]( · [◷](https://github.com/search?q=repo%3Amicrosoft%2Ftestfx+is%3Aissue+%22gh-aw-workflow-call-id%3A+microsoft%2Ftestfx%2Frepository-quality-improver%22&type=issues))
>
<details>
<summary><sub>Add this agentic workflow to your repo</sub></summary>

To install this agentic workflow, run

```
gh aw add githubnext/agentics/workflows/repository-quality-improver.md@main
```
</details>

> - [x] expires  on Aug 2, 2026, 10:38 PM UTC

### PR DESCRIPTION
**Pull request by Evangelink**
Fixes #10368.

## The report was partly stale

The issue says 0/21 packages set `PackageReadmeFile`. That is not the case: the root `Directory.Build.targets` already auto-wires a `PACKAGE.md` found next to a project, and 27 of them exist. Tasks 1, 2 and 3 of the issue are already done in the repo.

So I went looking for what is actually broken, and checked the nuget.org registration API rather than the sources.

## What is actually broken

**Three shipped packages publish NuGet's literal placeholder as their description.** Live on nuget.org today:

| Package | Published description |
| --- | --- |
| `Microsoft.Testing.Extensions.TrxReport.Abstractions` 2.3.3 | `Package Description` |
| `Microsoft.Testing.Platform.AI` 1.0.0-alpha | `Package Description` |
| `Microsoft.Testing.Extensions.AzureFoundry` 1.0.0-alpha | `Package Description` |

The root cause is a nasty one. NuGet's `Pack.targets` fills an unset `$(PackageDescription)` during **evaluation**:

```xml
<PackageDescription Condition="'$(PackageDescription)'==''">$(Description)</PackageDescription>
<PackageDescription Condition="'$(PackageDescription)'==''">Package Description</PackageDescription>
```

Arcade *does* have a `PackageDescription must be specified` check, but it lives in a **target**, so by the time it runs the placeholder is already there and the condition is never true. The check can never fire, and the gap ships silently.

**Every description led with boilerplate.** All 15 Microsoft.Testing.Platform packages opened with the same sentence about the product before saying what the package actually does — and nuget.org truncates the description in search results, which is exactly where that costs the most. Two of them (`HotReload`, `PackagedApp`) had also drifted to a *different* generic opening.

**The MSTest descriptions leak the .csproj indentation into the published text.** MSBuild trims a property value but not the indentation of its inner lines, so an inline multi-line `<PackageDescription>` publishes as `"MSTest is Microsoft supported Test Framework.\n\n      This package includes …"`.

## What this PR does

- Gives the three placeholder packages a real description.
- **Flips every description to lead with what the package does**, closing with the shared product sentence. This is Task 5 of the issue.
- Replaces the 15 duplicated copies of that shared sentence with `$(CommonProductDescription)`, referenced inline from an ordinary `<PackageDescription>` exactly the way projects already reference `$(CommonPackageTags)`. Nothing is hidden: each project still shows and greps its own complete description. (MSBuild expands property references inside CDATA, so the multi-paragraph descriptions keep their readable layout — and the .csproj indentation stays out of the published text.)
- Adds a `_ValidatePackageMetadata` guard that fails `pack` when a packable project has no description or no README. It keys off `$(_HasAuthoredPackageDescription)`, captured before NuGet's defaults land, and also matches the placeholder text as a second line of defence.
- Adds `PackageMetadataCompletenessTests`, modelled on the existing `PackageDependencyUpperBoundTests`, which inspects the produced `.nupkg` files for description, README and icon so the guard cannot silently stop firing.
- Documents the convention in `.github/copilot-instructions.md`.

## Published text that changes

Worth calling out, since these render on nuget.org at the next release:

- **Every package now leads with its specific purpose** instead of the shared product sentence, which moves to the end.
- `HotReload` and `PackagedApp` had drifted to a different generic sentence and now use the standard one like their siblings.
- `AzureDevOpsReport`: `a Azure DevOps` → `an Azure DevOps`.
- The MSTest packages lose the stray indentation described above.

No package-specific prose was rewritten — the existing sentences are preserved verbatim, only reordered.

## Note for whoever adds the next package

`pack` now fails outright for a packable project without a description or a `PACKAGE.md`, including `IsShipping=false` ones. That strictness is deliberate: `Microsoft.Testing.Extensions.AzureFoundry` is non-shipping and still reached nuget.org with the placeholder. The error text is self-documenting and points at the property to set.

## Verification

- `.\build.cmd -pack -c Debug`: 0 warnings, 0 errors.
- All 27 produced `.nupkg` files verified to carry a correct description, an embedded `PACKAGE.md` and an embedded icon.
- Both guard branches verified to fire (removing the description entirely; setting it explicitly to the placeholder text), and verified **not** to fire for a project that supplies its description via `$(Description)`.
- The new tests were mutation-tested with synthetic `.nupkg` files covering every violation branch, plus the stale-package and path-separator edge cases.

### CODE REVIEW

**Review by 0101 — APPROVED** [review]

**Review by 0101 — APPROVED** [review]

File: Directory.Build.targets [hunk 1]
```diff
@@ -15,11 +15,64 @@
...
+    Note this deliberately also covers IsShipping=false packages: Microsoft.Testing.Extensions.AzureFoundry
+    is non-shipping and still reached nuget.org with the placeholder.
+  -->
+  <Target Name="_ValidatePackageMetadata"
+          BeforeTargets="GenerateNuspec"
+          Condition=" '$(IsPackable)' == 'true' ">
+    <Error Condition=" '$(_HasAuthoredPackageDescription)' != 'true' or '$(PackageDescription)' == 'Package Description' "
+           Text="$(MSBuildProjectName) is packable but has no package description, so it would be published to nuget.org with NuGet's 'Package Description' placeholder. Set PackageDescriptionDetail (appended to the PackageDescriptionPreamble of the containing product area) or PackageDescription." />
+    <Error Condition=" '$(PackageReadmeFile)' == '' "
+           Text="$(MSBuildProjectName) is packable but has no package README, so its nuget.org page would show no rendered documentation. Add a PACKAGE.md next to the project file (it is picked up automatically) or set PackageReadmeFile." />
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply Directory.Build.targets:68]: Good catch, accepted - you are right that deleting this target or either `<Error>` left the produced packages unchanged and every test still green, so the enforcement could regress silently. Added `PackageMetadataGuardTests`, which packs a throwaway project omitting the description and the README in turn plus a valid control, asserting pack fails with the expected error and writes no package; the asset lives under `artifacts` so it inherits the repo Directory.Build files that put the guard in scope, and packs to its own output directory so a malformed package can never land in `artifacts/packages`. Mutation-tested: disabling either `<Error>` fails exactly the corresponding test while the control still passes.

File: test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/PackageMetadataCompletenessTests.cs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+    private const string NuGetPlaceholderDescription = "Package Description";
+
+    public TestContext TestContext { get; set; } = null!;
+
+    /// <summary>
+    /// Every produced package must carry a real description: nuget.org renders it on the package page and uses
+    /// it for search, so shipping NuGet's placeholder makes the package effectively undocumented.
+    /// </summary>
+    [TestMethod]
+    public void ProducedPackages_HaveARealDescription()
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/PackageMetadataCompletenessTests.cs]: Accepted - no produced package trips `HasIndentedLine`, so on its own that test could not tell a working detector from one that never reports anything. I extracted the per-package checks into a `PackageMetadataInspector` pure function so the real-package test and the new focused tests exercise the same code rather than a reimplementation, then added `PackageMetadataInspectorTests` covering the indented continuation line, the placeholder, an empty description, and both metadata files missing or declared-but-absent, plus a correct multi-paragraph description that must NOT be reported so the check cannot drift into false positives. Verified by making `HasIndentedLine` return false, which fails the new test.

File: test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/PackageMetadataCompletenessTests.cs [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+        List<string> violations = [];
+
+        foreach (ProducedPackage package in EnumerateProducedPackages())
+        {
+            checkedPackages.Add(package.FileName);
+            violations.AddRange(PackageMetadataInspector.InspectPackage(package.FileName, package.Nuspec, package.Entries));
+        }
+
+        AssertSomethingWasPacked(checkedPackages);
+        Assert.IsEmpty(
```
[the remark this answers is not in the document]
Comment (Evangelink) [reply test/IntegrationTests/Microsoft.Testing.Platform.Acceptance.IntegrationTests/PackageMetadataCompletenessTests.cs:55]: Agreed, and no code change needed here - as your own "how to improve" note says, this is what the two-class design is for. I confirmed it empirically rather than assuming: stubbing `InspectPackage` to always return `[]` fails 7 of the 12 tests, all from `PackageMetadataInspectorTests`, so that mutation is caught loudly at the suite level even though this particular test would stay green in isolation. The class remarks already document that split, so I am resolving this as by-design.


> [!NOTE]
> Auto-resolved by the [GitHub Copilot app](https://github.com/github/app).

**Review by 0101 — APPROVED** [review]
